### PR TITLE
[c++][NOMERGE] Refactor C++ Code

### PR DESCRIPTION
**Pull request by nguyenv**
**Issue and/or context:**
#1240

**Changes:**

**Notes for Reviewer:**
WIP

### CODE REVIEW

File: libtiledbsoma/src/cpp_api/soma_measurement.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+   private:
+    //===================================================================
+    //= private non-static
+    //===================================================================
+
+    // TileDB context
+    std::shared_ptr<Context> ctx_;
+
+    // SOMAGroup
+    std::unique_ptr<SOMAGroup> group_;
```
Comment (beroy) [inline libtiledbsoma/src/cpp_api/soma_measurement.h:62]: @hat's group here?

File: libtiledbsoma/src/cpp_api/soma_measurement.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    // A collection of matrices, each containing measured feature values
+    SOMACollection X_;
+
+    // A collection of dense matrices containing annotations of each obs row
+    SOMACollection obsm_;
+
+    // A collection of sparse matrices containing pairwise annotations of each
+    // obs row
+    SOMACollection obsp_;
```
Comment (beroy) [inline libtiledbsoma/src/cpp_api/soma_measurement.h:75]: Shouldn't ```varm_```  and ```var_p``` be DF too?

File: libtiledbsoma/src/cpp_api/soma_object.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ *   This file defines the SOMAObject class.
```
Comment (beroy) [inline libtiledbsoma/src/cpp_api/soma_object.h:30]: Is it the base class for all SOMA objects?

**Comment by johnkerl** [comment]
@nguyenv This PR is in draft, and with `[WIP]` in its title -- and I respect that -- I don't think you're ready for review yet so I'll hold off.

Looking forward though -- please let me know your plans for this PR:

* If this is "one big tree to pluck apples from" that you'll later carve into smaller PRs, great!
* If this is "one big PR as cloud backup in case your laptop crashes", also great!
* But if your plan moving forward is to undraft this PR more or less as-is -- I'd ask you to split it up
  * In general, several smaller PRs are easier on everyone
  * Really big PRs that are "all green" -- all new code -- are still not my first choice, but are a little easier to read
  * But really big PRs that are "red/green mix" -- old code, new code, refactors, etc -- this is too much for the reviewers.

**Comment by nguyenv** [comment]
I've decided that this is going to be the first bullet point where I'll carve this into smaller PRs. Yesterday I spent quite a bit of time refactoring the code a bit to support that. Only the headers in `cpp_api` are meant to be public but if we merged it in like that right away, we would need to completely refactor the R and Python APIs to reflect that. For now, I have all the headers public. This way we can start merging in the code little by little without breaking the code. 

This draft PR at the moment contains:

#1247 
#1248
#1122
#1256
#1257
#1258 

I'll definitely split those out into different PRs.

File: libtiledbsoma/src/soma/soma_array.h [hunk 4]
```diff
@@ -301,43 +316,133 @@ class SOMAArrayReader {
...
+     *
+     * @param key The key of the metadata item to be retrieved. UTF-8 encodings
+     *     are acceptable.
+     * @param value_type The datatype of the value associated with the key (if
+     * any).
+     * @return true if the key exists, else false.
+     * @note If the key does not exist, then `value_type` will not be modified.
+     */
+    bool has_metadata(const std::string&, tiledb_datatype_t*);
+
```
[the remark this answers is not in the document]
Comment (nguyenv) [reply libtiledbsoma/src/soma/soma_array.h:460]: Yeah you're right. I will add the argument names into the header files.

File: libtiledbsoma/src/cpp_api/soma_object.h (hunk 3 above)
Comment (nguyenv) [reply libtiledbsoma/src/cpp_api/soma_object.h:30]: This is the base class for the public SOMA objects so not `SOMAArray` or `SOMAGroup`. But includes `SOMACollection`, `SOMAExperiment`, `SOMAMeasurement`, `SOMA{Sparse,Dense}NdArray`, and `SOMADataFrame`.

File: libtiledbsoma/src/cpp_api/soma_experiment.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+   private:
+    //===================================================================
+    //= private non-static
+    //===================================================================
+
+    // TileDB context
+    std::shared_ptr<Context> ctx_;
+
+    // SOMAGroup
+    std::unique_ptr<SOMAGroup> group_;
```
[the remark this answers is not in the document]
Comment (nguyenv) [reply libtiledbsoma/src/cpp_api/soma_experiment.h]: A `SOMAGroup` is essentially just a TileDB `Group` with some SOMA specific methods like `member_name_to_uri`, etc. I'll also go back in and replace the usage of raw pointers with smart pointers. `SOMAGroup` is an internal structure. `SOMACollection`  is implemented with `SOMAGroup`. `SOMAExperiment` and `SOMAMeasurement` inherit from `SOMACollection`.

File: libtiledbsoma/src/cpp_api/soma_measurement.h (hunk 1 above)
Comment (johnkerl) [reply libtiledbsoma/src/cpp_api/soma_measurement.h:69]: @beroy same as above

File: libtiledbsoma/src/cpp_api/soma_measurement.h (hunk 2 above)
Comment (nguyenv) [reply libtiledbsoma/src/cpp_api/soma_measurement.h:82]: I see these labelled as `SOMACollection` in [here](https://github.com/single-cell-data/SOMA/blob/main/abstract_specification.md#somaexperiment-and-somameasurement
).

File: libtiledbsoma/src/cpp_api/soma_dense_ndarray.h [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+     * @param ctx TileDB context
+     * @param uri URI of the array
+     * @param name Name of the array
+     * @param column_names Columns to read
+     * @param batch_size Batch size
+     * @param result_order Result order
+     * @param timestamp Timestamp
+     * @return std::unique_ptr<SOMADenseNDArray> SOMADenseNDArray
+     */
+    static std::unique_ptr<SOMAArray> open(
```
[the remark this answers is not in the document]
Comment (nguyenv) [reply libtiledbsoma/src/cpp_api/soma_dense_ndarray.h]: One takes a `std::shared_ptr<Context>` and the other takes a `std::map<std::string, std::string> platform_config`. It's something I copy and pasted from the already existing `SOMAArrayReader` but would be good to keep both options to pass either a context or config.

File: libtiledbsoma/src/soma/soma_array.h [hunk 7]
```diff
@@ -301,43 +316,133 @@ class SOMAArrayReader {
...
+     * @param key The metadata key.
+     * @param value_type The datatype of the value.
+     * @param value_num The value may consist of more than one items of the
+     *     same datatype. This argument indicates the number of items in the
+     *     value component of the metadata. Keys with empty values are indicated
+     *     by value_num == 1 and value == NULL.
+     * @param value The metadata value in binary form.
+     */
+    void get_metadata_from_index(
+        uint64_t, std::string*, tiledb_datatype_t*, uint32_t*, const void**);
```
[the remark this answers is not in the document]
Comment (nguyenv) [reply libtiledbsoma/src/soma/soma_array.h:446]: I'll change them to smart pointers. I kept the raw pointers from the TileDB code but should be more midnful about lifetime.

File: libtiledbsoma/src/soma/soma_array.h [hunk 8]
```diff
@@ -301,43 +316,133 @@ class SOMAArrayReader {
...
+    void get_metadata(
+        const std::string&, tiledb_datatype_t*, uint32_t*, const void**);
+
+    /**
+     * It gets a metadata item from an open array using an index.
+     * The array must be opened in READ mode, otherwise the function will
+     * error out.
+     *
+     * @param index The index used to get the metadata.
+     * @param key The metadata key.
```
[the remark this answers is not in the document]
Comment (nguyenv) [reply libtiledbsoma/src/soma/soma_array.h:437]: I'll improve the documentation for this. We only use the index to get the metadata here. We don't know what the key is beforehand. This method will set the paramenter to the metadata key.

File: libtiledbsoma/src/cpp_api/soma_experiment.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (beroy) [reply libtiledbsoma/src/cpp_api/soma_experiment.h]: Thanks a lot for clarification @nguyenv and @johnkerl

File: libtiledbsoma/src/cpp_api/soma_object.h (hunk 3 above)
Comment (beroy) [reply libtiledbsoma/src/cpp_api/soma_object.h:30]: Thanks. I would make that clarification in the comments.

**Comment by nguyenv** [comment]
All the code in this PR has now been separated out and expanded in multiple PRs and can now be closed. Refer to https://github.com/single-cell-data/TileDB-SOMA/issues/1240 to reference the refactored SOMA C++ code.

**Pull request closed without merging**